bugbug/models/duplicate.py

```diff
@@ -0,0 +1,135 @@
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import LabelEncoder
+from sklearn.svm import LinearSVC
+
+from bugbug import bug_features, bugzilla, feature_cleanup
+from bugbug.model import BugCoupleModel
+
+
+class LinearSVCWithLabelEncoding(CalibratedClassifierCV):
+    def __init__(self, clf):
+        super().__init__(clf)
+        self._le = LabelEncoder()
+
+    def fit(self, X, y):
+        super().fit(X, y)
+        self._le.fit(y)
+
+
+class DuplicateModel(BugCoupleModel):
+    def __init__(self, lemmatization=False):
+        BugCoupleModel.__init__(self, lemmatization)
+
+        self.sampler = BorderlineSMOTE(random_state=0)
+        self.calculate_importance = False
+
+        cleanup_functions = [
+            feature_cleanup.responses(),
+            feature_cleanup.hex(),
+            feature_cleanup.dll(),
+            feature_cleanup.fileref(),
+            feature_cleanup.url(),
+            feature_cleanup.synonyms(),
+            feature_cleanup.crash(),
+        ]
+
+        self.extraction_pipeline = Pipeline(
+            [
+                ("bug_extractor", bug_features.BugExtractor([], cleanup_functions)),
+                (
+                    "union",
+                    ColumnTransformer([("text", self.text_vectorizer(), "text")]),
+                ),
+            ]
+        )
+
+        self.clf = LinearSVCWithLabelEncoding(LinearSVC())
+
+    def get_labels(self):
+        bugs = []
+
+        for bug_data in islice(bugzilla.get_bugs(), 0, 20000):
+            bugs.append(bug_data["id"])
+
+        classes = {}
+        # Only store id's of bugs that have duplicate or are duplicate
+        duplicate_ids = []
+
+        # Store all remaining ids
+        all_ids = []
+
+        count = 4500
+        for bug_data in bugzilla.get_bugs():
+
+            if bug_data["duplicates"] != []:
+                duplicate_ids.append(bug_data["id"])
+                for duplicate_bug in bug_data["duplicates"]:
+                    if duplicate_bug in bugs:
+                        duplicate_ids.append(duplicate_bug)
+                        classes[(bug_data["id"], duplicate_bug)] = 1
+                        # classes[(bug_data["id"], duplicate_bug)] = {
+                        #    "label": 1,
+                        #    "data": (bugs[bug_data["id"]], bugs[duplicate_bug]),
+                        # }
+                        count -= 1
+                    if not count:
+                        break
+            if not count:
+                break
+
+        for bug in bugs:
+            if bug not in duplicate_ids:
+                all_ids.append(bug)
+
+        le = len(classes)
+        print("Number of purely duplicate labels are: ", le)
+
+        all_ids = set(all_ids)
+
+        duplicate_ids = set(duplicate_ids)
+
+        # When the bug has no duplicates, we create dup-nondup labels = 0
+        count = 2500
+        for key in duplicate_ids:
+            for key2 in all_ids:
```

Comment:
Remember that iterating sets is really slow compared to iterating lists, so you should use lists here